### PR TITLE
Serf encryption

### PR DESCRIPTION
**Pull request by Gerrrr**
Following the discussion at #469, this is the first PR which enables serf encryption.
Most of the code is taken from Consul codebase.

User-visible changes:

* new server-specific CLI option for `nomad agent` - `-encrypt`
* new server-specific configuration option `encrypt`
* new CLI option `nomad keygen` which generates new encryption keys

Example configuration:

`server.hcl`
```hcl
data_dir = "/var/lib/nomad"
server {
    encrypt = "pd3LCI6Um4o6KrxEva7ylw=="
    enabled = true
    bootstrap_expect = 3
    start_join = ["172.28.128.3", "172.28.128.4", "172.28.128.5"]
}
```

`server01.hcl`
```hcl
bind_addr = "172.28.128.3"
```

`server02.hcl`
```hcl
bind_addr = "172.28.128.4"
```

`server03.hcl`
```hcl
bind_addr = "172.28.128.5"
```

### CODE REVIEW

File: command/agent/agent.go [hunk 1]
```diff
@@ -356,6 +357,29 @@ func (a *Agent) clientConfig() (*clientconfig.Config, error) {
 	return conf, nil
 }
 
+// setupKeyring is used to initialize and load keyring during server startup
+func (a *Agent) setupKeyring(config *nomad.Config) error {
+	file := filepath.Join(a.config.DataDir, serfWANKeyring)
```
Comment (dadgar) [inline command/agent/agent.go]: `s/serfWANKeyring/serfLANKeyring`

File: command/agent/command.go [hunk 2]
```diff
@@ -212,6 +213,13 @@ func (c *Command) readConfig() *Config {
 		}
 	}
 
+	if config.Server.Enabled && config.Server.EncryptKey != "" {
+		if _, err := config.EncryptBytes(); err != nil {
```
Comment (dadgar) [inline command/agent/command.go]: Should this go in setupKeyring?

File: command/agent/keyring.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	"os"
+	"path/filepath"
+
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/memberlist"
+	"github.com/hashicorp/serf/serf"
+)
+
+const (
+	serfWANKeyring = "serf/remote.keyring"
```
Comment (dadgar) [inline command/agent/keyring.go]: Put a comment on this

File: command/keygen.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	return 0
+}
+
+func (c *KeygenCommand) Synopsis() string {
+	return "Generates a new encryption key"
+}
+
+func (c *KeygenCommand) Help() string {
+	helpText := `
+Usage: consul keygen
```
Comment (dadgar) [inline command/keygen.go]: Consul?

Add 
```
General Options:

  ` + generalOptionsUsage()
```

File: command/agent/keyring.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+// keyringProcess is used to abstract away the semantic similarities in
+// performing various operations on the encryption keyring.
+func (a *Agent) keyringProcess(args *structs.KeyringRequest) (*structs.KeyringResponses, error) {
+	var reply structs.KeyringResponses
+	if a.server == nil {
+		return nil, fmt.Errorf("keyring operations must run against a server node")
+	}
+	if err := a.RPC("Internal.KeyringOperation", args, &reply); err != nil {
```
Comment (dadgar) [inline command/agent/keyring.go]: This endpoint does not exist. You will have to expose a method on the server to get the serf key manager

File: command/agent/keyring_test.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+	c = agent2.config.NomadConfig
+
+	if c.SerfConfig.KeyringFile == "" {
+		t.Fatalf("should have keyring file")
+	}
+	if c.SerfConfig.MemberlistConfig.Keyring == nil {
+		t.Fatalf("keyring should be loaded")
+	}
+}
```
Comment (dadgar) [inline command/agent/keyring_test.go]: Add tests for all operations

File: command/agent/keyring_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+	if c.SerfConfig.KeyringFile != "" {
+		t.Fatalf("bad: %#v", c.SerfConfig.KeyringFile)
+	}
+	if c.SerfConfig.MemberlistConfig.Keyring != nil {
+		t.Fatalf("keyring should not be loaded")
+	}
+
+	// Server should auto-load WAN keyring file
+	dir2, agent2 := makeAgentKeyring(t, nil, key)
```
Comment (dadgar) [inline command/agent/keyring_test.go]: command/agent/keyring_test.go:26: undefined: makeAgentKeyring

**Comment by dadgar** [comment]
Please do not rely on the Travis green checkmark for now. For some reason the builds return passing even with non-zero exit codes. Worthing with Travis to fix.

This branch does not build properly

**Comment by kevincox** [comment]
It built fine for me.

**Comment by kevincox** [comment]
So it seems to be working for me. I'm going to try running a job but status commands and the like are working.

Also IIUC this only encrypts serf traffic? Do you have plans for HTTP and RPC encryption in this PR or will that be coming at a different time?

File: command/agent/agent.go (hunk 1 above)
Comment (Gerrrr) [inline command/agent/agent.go]: https://www.nomadproject.io/docs/internals/gossip.html here I found that all the Nomad servers participate in a single global WAN pool, and also `SerfConfig` uses WAN-specific defaults (https://github.com/hashicorp/nomad/blob/e0a64086764f3021068f637a6c5c4bbf51e2cce8/nomad/config.go#L259). Would you mind briefly explaining this change request?

**Comment by Gerrrr** [comment]
@kevincox yes, this PR only adds Serf traffic encryption. I am also working on the PR for RPC traffic encryption, probably will publish it this week.

**Comment by dadgar** [comment]
@kevincox The tests don't build, I should have clarified

File: command/agent/agent.go (hunk 1 above)
Comment (dadgar) [inline command/agent/agent.go]: You are totally right! Sorry about that! Will delete comment

**Comment by kevincox** [comment]
@dadgar Ah, I see. I guess I just built the binary.

@Gerrrr Awesome. I would be ready to start running it "in production" once RPC is also done. Even if the HTTP isn't protected. I'll open an issue for the HTTP as there should probably be a discussion on the best way to do that.

**Comment by Gerrrr** [comment]
@dadgar Can you please take another look at the PR?

File: command/keygen.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"strings"
+)
+
+// KeygenCommand is a Command implementation that generates an encryption
+// key for use in `consul agent`.
```
Comment (dadgar) [inline command/keygen.go]: nomad

File: nomad/keyring_operation_endpoint.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
+package nomad
+
+import (
+	"errors"
+
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/serf/serf"
+)
+
+type KeyringOperation struct {
```
Comment (dadgar) [inline nomad/keyring_operation_endpoint.go]: Can you name this `Keyring`

File: nomad/keyring_operation_endpoint.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
+package nomad
+
+import (
+	"errors"
+
+	"github.com/hashicorp/consul/consul/structs"
```
Comment (dadgar) [inline nomad/keyring_operation_endpoint.go]: You need to be using `nomad/structs`

File: nomad/keyring_operation_endpoint_test.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/net-rpc-msgpackrpc"
+	"github.com/hashicorp/nomad/testutil"
+)
+
+const (
+	key = "H1dfkSZOVnP/JUnaBfTzXg=="
+)
+
+func TestKeyringOperationEndpoint_SingleNodeCluster(t *testing.T) {
```
Comment (dadgar) [inline nomad/keyring_operation_endpoint_test.go]: Test use/install/remove

File: commands.go [hunk 12]
```diff
@@ -79,6 +79,11 @@ func Commands(metaPtr *command.Meta) map[string]cli.CommandFactory {
 				Meta: meta,
 			}, nil
 		},
+		"keygen": func() (cli.Command, error) {
```
Comment (dadgar) [inline commands.go]: `keyring` command so we can manage these keys?

**Comment by dadgar** [comment]
Added some more comments. Before this can be merged we will have to also document the new config fields and flags and new commands in the `website/docs`

File: commands.go (hunk 12 above)
Comment (Gerrrr) [inline commands.go]: What internal interface do you want to have to perform the queries? In Consul there is an [rpc_client](https://github.com/hashicorp/consul/blob/master/command/keyring.go#L60), but it is also reused in other commands. I thought of adding an http endpoint, like it is done for other commands, but wanted to know your opinion first.

File: command/agent/config.go [hunk 13]
```diff
@@ -195,6 +196,9 @@ type ServerConfig struct {
 	// DataDir is the directory to store our state in
 	DataDir string `mapstructure:"data_dir"`
 
+	// Encryption key to use for the Serf communication
+	EncryptKey string `mapstructure:"encrypt" json:"-"`
```
Comment (dadgar) [inline command/agent/config.go]: `json:"-"`?

File: nomad/structs/structs.go [hunk 14]
```diff
@@ -3178,3 +3178,54 @@ func Encode(t MessageType, msg interface{}) ([]byte, error) {
...
+// It can be used for install, remove, or use key type operations.
+type KeyringRequest struct {
+	Operation  KeyringOp
+	Key        string
+	Datacenter string
+	Forwarded  bool
+	QueryOptions
+}
+
+func (r *KeyringRequest) RequestDatacenter() string {
```
Comment (dadgar) [inline nomad/structs/structs.go]: Remove?

File: nomad/structs/structs.go [hunk 15]
```diff
@@ -3178,3 +3178,54 @@ func Encode(t MessageType, msg interface{}) ([]byte, error) {
...
+
+// KeyringResponses holds multiple responses to keyring queries. Each
+// datacenter replies independently, and KeyringResponses is used as a
+// container for the set of all responses.
+type KeyringResponses struct {
+	Responses []*KeyringResponse
+	QueryMeta
+}
+
+func (r *KeyringResponses) Add(v interface{}) {
```
Comment (dadgar) [inline nomad/structs/structs.go]: Remove `Add` and `New`

**Comment by dadgar** [comment]
Hey Aleksandr,

Appreciate the work you have done so far. Within the next couple days we will have some bandwidth to take this PR over. I think it is pretty core feature and think it is best if we complete the rest of the work. Please just leave this PR open and we will branch from it.

Thanks,
Alex

**Comment by diptanu** [comment]
Closing in favor of #1791 

@Gerrrr Thanks for your hard work on this. Please review the new PR I have opened, and let us know what you think. We implemented a few things differently than how Consul has implemented this feature especially our API around key management is http based instead of Consul's RPC based interface.

**Pull request closed without merging**